src/libsyntax/mut_visit.rs

```diff
@@ -300,6 +288,38 @@ pub trait MutVisitor: Sized {
     fn visit_span(&mut self, _sp: &mut Span) {
         // Do nothing.
     }
+
+    fn flat_map_arg(&mut self, arg: Arg) -> SmallVec<[Arg; 1]> {
+        noop_flat_map_arg(arg, self)
+    }
+
+    fn flat_map_arm(&mut self, arm: Arm) -> SmallVec<[Arm; 1]> {
+        noop_flat_map_arm(arm, self)
+    }
+
+    fn flat_map_field(&mut self, f: Field) -> SmallVec<[Field; 1]> {
+        noop_flat_map_field(f, self)
+    }
+
+    fn flat_map_field_pattern(&mut self, fp: FieldPat) -> SmallVec<[FieldPat; 1]> {
+        noop_flat_map_field_pattern(fp, self)
+    }
+}
+
+pub fn noop_flat_map_field_pattern<T: MutVisitor>(
+    fp: FieldPat,
+    _vis: &mut T,
+) -> SmallVec<[FieldPat; 1]> {
+    let FieldPat {
+        attrs: _,
+        id: _,
+        ident: _,
+        is_shorthand: _,
+        pat: _,
+        span: _,
+        is_placeholder: _
+    } = fp;
```

Comment:
This function should visit all the nested things (`attrs`, `pat`, etc.) and `noop_visit_pat` should call it.